### DISCUSSION TITLE
How to convert a string with Hex value to Int in Berry?

### DISCUSSION THREAD

**Discussion by PhilipTS** [arendst/Tasmota#14380] [Q&A] [answered]
Hello,

I have a string that contains a HEX value and I want to convert it to an Integer.
e.x. "AB" -> 171

In other words i want to do the reverse of ...

import string
print("DEC 171 = HEX "+string.hex(171))
DEC 171 = HEX AB

Some help would be really appreciated.

Long Version
I have a TuyaMCU Thermostat and I connected it to a ESP32 DevKit.
I have got it fully working but with ugly solutions.
I ended up doing a Map like { 'AB':'171' } but I would like to do it properly, it's my "OCD".

**Comment by sfromis** [comment, marked as answer]
One way is via the bytes class.
`print(bytes("0abc").get(0,-2))`
`2748`
Not sure if there is a more elegant way, but the above can easily be wrapped in a function.

The bytes class is even designed to allow you to have a longer hex buffer, and pick out parts which you can get as a number, signed or unsigned, little or big-endian.

**Comment by PhilipTS** [comment]
@sfromis 

Thank you very much,  it looks way better that what I was preparing :)
Will study the bytes class to figure out how it works.

I really really like tasmota with berry.
Did i say that I really really like tasmota with berry?
Yes yes I said that i really really like tasmota with berry.
Like REALLY RELLY :)

```
import math

def hexStr2int(hexStr)
  var y
  var index
  var digit
  var maxIndex
  var sum = 0
  var digits = { '0':0 , '1':1 ,'2':2 ,'3':3 ,'4':4 ,'5':5 ,'6':6 ,'7':7 ,'8':8 ,'9':9 , 'A':10 ,'B':11 ,'C':12 ,'D':13 ,'E':14 ,'F':15 }

  maxIndex = size(hexStr) -1

  for i: 0 .. maxIndex
	y = maxIndex - i
    digit = int(digits.find(hexStr[i]))
    sum = sum + (digit * math.pow(16, y))
  end
  print(sum)
  return sum
end
```

**Reply by sfromis** [reply]
Yeah, Berry is the cherry on top :-)

**Comment by nccchirag** [comment]
the below may work for simple int converstions. .. which may not require size calculations like byte class
print(int(str("0x"+"320"))) # 800

Indeed berry truely amplifies tasmota 🤩, kudos to all the contributors 🙌🏻

**Reply by sfromis** [reply]
The same, but shorter:
`print(int("0x"+"320"))`
Of course, it does not work well for negative numbers.
